### PR TITLE
Suggest desugaring to return-position `impl Future` when an `async fn` in trait fails an auto trait bound

### PR DESCRIPTION
**Pull request by compiler-errors**
First commit allows us to store the span of the `async` keyword in HIR.

Second commit implements a suggestion to desugar an `async fn` to a return-position `impl Future` in trait to slightly improve the `Send` situation being discussed in #115822.

This suggestion is only made when `#![feature(return_type_notation)]` is not enabled -- if it is, we should instead suggest an appropriate where-clause bound.

### CODE REVIEW

**Comment by compiler-errors** [comment]
Because of the changes to HIR

@bors try @rust-timer queue

**Comment by compiler-errors** [comment]
Of course because I'm developing on aarch64, I didn't see any of the type size changes.

@rustbot author

File: compiler/rustc_trait_selection/src/traits/error_reporting/suggestions.rs
Comment (estebank) [inline compiler/rustc_trait_selection/src/traits/error_reporting/suggestions.rs:4067]: Should we maybe *mention* that the problem lies in the trait definition so they know they have to bother the crate maintainer? Particularly thinking of the case of workspaces where the crate isn't *local*, but still under the dev's control.

File: compiler/rustc_trait_selection/src/traits/error_reporting/suggestions.rs
Comment (estebank) [inline compiler/rustc_trait_selection/src/traits/error_reporting/suggestions.rs:4103]: should we at least log these? If only we had metrics...

File: compiler/rustc_trait_selection/src/traits/error_reporting/suggestions.rs
Comment (estebank) [inline compiler/rustc_trait_selection/src/traits/error_reporting/suggestions.rs:4140]: Don't block on this, but it'd be nice to put the `async {` and closing `}` at the same indentation level as the rest of the body, on their own line. The suggestion for `test2` below ending up all in one line won't be too pretty.

**Review by estebank — APPROVED** [review]

File: compiler/rustc_trait_selection/src/traits/error_reporting/suggestions.rs
Comment (compiler-errors) [reply compiler/rustc_trait_selection/src/traits/error_reporting/suggestions.rs:4140]: I'm gonna let rustfmt do this. I still don't think this looks better:

```rust
    fn foo() -> impl Future + Send {
        async {
        the();
        body();
        is();
        here();
        }
    }
```

than this:

```rust
    fn foo() -> impl Future + Send { async {
        the();
        body();
        is();
        here();
    } }
```

Both are iffy, and rely on us having to use `rustfmt` anyways 😅

File: compiler/rustc_trait_selection/src/traits/error_reporting/suggestions.rs
Comment (compiler-errors) [reply compiler/rustc_trait_selection/src/traits/error_reporting/suggestions.rs:4067]: Yeah, I can do that. I can at least point to the async fn definition if we have a span for that.

File: compiler/rustc_trait_selection/src/traits/error_reporting/suggestions.rs
Comment (estebank) [reply compiler/rustc_trait_selection/src/traits/error_reporting/suggestions.rs:4140]: Yeah, I guess it was particularly ugly because it was a single line body. rustfmt ftw

**Comment by nnethercote** [comment]
> Of course because I'm developing on aarch64, I didn't see any of the type size changes.

It might be worth changing how that condition works, so that it hits for Mac M1 users.

**Comment by compiler-errors** [comment]
@nnethercote: unfortunately, can't do it the "simple" way by just enabling the `static_assert_size` checks on aarch64, since the sizes of items change between x86_64 and aarch64 -- @JakobDegen told me it was probably due to types containing `u128`s. and I think that maintaining two sets of type sizes is a lot of work 😅

**Comment by nnethercote** [comment]
I wonder what fraction of the existing size assertions have different results on aarch64 vs x86-64.

**Pull request merged**
